jxl/src/render/stages/upsample.rs

```diff
@@ -3,135 +3,82 @@
-        write!(
-            f,
-            "chroma upsample of channel {}, horizontally",
-            self.channel
-        )
-    }
-}
-
-impl RenderPipelineStage for HorizontalChromaUpsample {
-    type Type = RenderPipelineInOutStage<f32, f32, 1, 0, 1, 0>;
-
-    fn uses_channel(&self, c: usize) -> bool {
-        c == self.channel
-    }
-
-    fn process_row_chunk(
-        &mut self,
-        _position: (usize, usize),
-        xsize: usize,
-        row: &mut [(&[&[f32]], &mut [&mut [f32]])],
-    ) {
-        let (input, output) = &mut row[0];
-        for i in 0..xsize {
-            let scaled_cur = input[0][i + 1] * 0.75;
-            let prev = input[0][i];
-            let next = input[0][i + 2];
-            let left = 0.25 * prev + scaled_cur;
-            let right = 0.25 * next + scaled_cur;
-            output[0][2 * i] = left;
-            output[0][2 * i + 1] = right;
+impl Upsample2x {
+    #[allow(dead_code)]
+    pub fn new(ups_factors: &CustomTransformData, channel: usize) -> Upsample2x {
+        let weights = ups_factors.weights2;
+        let n = 1;
+        let mut kernel = [[[[0.0; 5]; 5]; 2]; 2];
+        for i in 0..5 * n {
+            for j in 0..5 * n {
+                let y = i.min(j);
+                let x = i.max(j);
+                let y = y as isize;
+                let x = x as isize;
+                let n = n as isize;
+                let index = (5 * n * y - y * (y - 1) / 2 + x - y) as usize;
+                // Filling in the top left corner from the weights
+                kernel[j / 5][i / 5][j % 5][i % 5] = weights[index];
+                // Mirroring to get the rest of the kernel.
+                kernel[(n as usize) + j / 5][i / 5][4 - (j % 5)][i % 5] = weights[index];
+                kernel[j / 5][(n as usize) + i / 5][j % 5][4 - (i % 5)] = weights[index];
+                kernel[(n as usize) + j / 5][(n as usize) + i / 5][4 - (j % 5)][4 - (i % 5)] =
+                    weights[index];
+            }
         }
+        Upsample2x { kernel, channel }
     }
 }
 
-pub struct VerticalChromaUpsample {
-    channel: usize,
-}
-
-impl VerticalChromaUpsample {
-    pub fn new(channel: usize) -> VerticalChromaUpsample {
-        VerticalChromaUpsample { channel }
-    }
-}
-
-impl std::fmt::Display for VerticalChromaUpsample {
+impl std::fmt::Display for Upsample2x {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "chroma upsample of channel {}, vertically", self.channel)
+        write!(f, "2x2 upsampling of channel {}", self.channel)
     }
 }
 
-impl RenderPipelineStage for VerticalChromaUpsample {
-    type Type = RenderPipelineInOutStage<f32, f32, 0, 1, 0, 1>;
+impl RenderPipelineStage for Upsample2x {
+    type Type = RenderPipelineInOutStage<f32, f32, 2, 2, 1, 1>;
 
     fn uses_channel(&self, c: usize) -> bool {
         c == self.channel
     }
-
+    // Takes a 5x5 area and upsamples it to 10x10 using the kernel.
```

Comment:
Not quite a 5x5 area, right?
(also, I assume this was meant to be a doc comment?)

Reply:
made it a doc comment and made it more precise